cmd/soroban-rpc/internal/feewindow/feewindow.go

```diff
@@ -0,0 +1,191 @@
+	return &FeeWindow{
+		feesPerLedger: window,
+	}
+}
+
+func (fw *FeeWindow) AppendLedgerFees(fees ledgerbucketwindow.LedgerBucket[[]uint64]) error {
+	fw.lock.Lock()
+	defer fw.lock.Unlock()
+	_, err := fw.feesPerLedger.Append(fees)
+	if err != nil {
+		return err
+	}
+
+	var allFees []uint64
+	for i := uint32(0); i < fw.feesPerLedger.Len(); i++ {
+		allFees = append(allFees, fw.feesPerLedger.Get(i).BucketContent...)
+	}
+	fw.distribution = computeFeeDistribution(allFees, fw.feesPerLedger.Len())
+
+	return nil
+}
+
+func computeFeeDistribution(fees []uint64, ledgerCount uint32) FeeDistribution {
+	if len(fees) == 0 {
+		return FeeDistribution{}
+	}
+	slices.Sort(fees)
+	mode := fees[0]
+	lastVal := fees[0]
+	maxRepetitions := 0
+	localRepetitions := 0
+	for i := 1; i < len(fees); i++ {
+		if fees[i] == lastVal {
+			localRepetitions += 1
+			continue
+		}
+
+		// new cluster of values
+
+		if localRepetitions > maxRepetitions {
+			maxRepetitions = localRepetitions
+			mode = lastVal
+		}
+		lastVal = fees[i]
+		localRepetitions = 0
+	}
+
+	if localRepetitions > maxRepetitions {
+		// the last cluster of values was the longest
+		mode = fees[len(fees)-1]
+	}
+
+	count := len(fees)
+	// nearest-rank percentile
+	percentile := func(p uint64) uint64 {
+		// ceiling(p*count/100)
+		kth := ((p * uint64(count)) + 100 - 1) / 100
+		return fees[kth-1]
+	}
+	return FeeDistribution{
+		Max:         fees[len(fees)-1],
+		Min:         fees[0],
+		Mode:        mode,
+		P10:         percentile(10),
+		P20:         percentile(20),
+		P30:         percentile(30),
+		P40:         percentile(40),
+		P50:         percentile(50),
+		P60:         percentile(60),
+		P70:         percentile(70),
+		P80:         percentile(80),
+		P90:         percentile(90),
+		P95:         percentile(95),
+		P99:         percentile(99),
+		FeeCount:    uint32(count),
+		LedgerCount: ledgerCount,
+	}
+}
+
+func (fw *FeeWindow) GetFeeDistribution() FeeDistribution {
+	fw.lock.RLock()
+	defer fw.lock.RUnlock()
+	return fw.distribution
+}
+
+type FeeWindows struct {
+	SorobanInclusionFeeWindow *FeeWindow
+	ClassicFeeWindow          *FeeWindow
+	networkPassPhrase         string
+}
+
+func NewFeeWindows(classicRetention uint32, sorobanRetetion uint32, networkPassPhrase string) *FeeWindows {
+	return &FeeWindows{
+		SorobanInclusionFeeWindow: NewFeeWindow(sorobanRetetion),
+		ClassicFeeWindow:          NewFeeWindow(classicRetention),
+		networkPassPhrase:         networkPassPhrase,
+	}
+}
+
+func (fw *FeeWindows) IngestFees(meta xdr.LedgerCloseMeta) error {
+	reader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(fw.networkPassPhrase, meta)
+	if err != nil {
+		return err
+	}
+	var sorobanInclusionFees []uint64
+	var classicFees []uint64
+	for {
+		tx, err := reader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		feeCharged := uint64(tx.Result.Result.FeeCharged)
```

Comment:
it looks like the horizon endpoint includes 2 distributions one for FeeCharged and another for MaxFee (which is the maximum fee the account is willing to pay to submit their transaction):

https://horizon.stellar.org/fee_stats

Here we only include the distribution on FeeCharged. Is that intentional or should we also include both distributions in the soroban-rpc endpoint?

Reply:
CC @janewang

Reply:
The distribution makes sense to show is the actual feeCharged.